Conditional magic methods cause false positives

```
$ cat foo.py
class Foo:
    if x:
        def __init__():
            pass

$ flake8 foo.py --select=N
foo.py:3:13: N802 function name should be lowercase

$ flake8 --version
3.5.0 (mccabe: 0.6.1, naming: 0.4.1, pycodestyle: 2.3.1, pyflakes: 1.6.0) CPython 3.5.3 on Linux
```

--- a/src/pep8ext_naming.py
+++ b/src/pep8ext_naming.py
@@ -19,6 +19,11 @@ UPPERCASE_REGEX = re.compile(r'[_A-Z][_A-Z0-9]*$')
 MIXEDCASE_REGEX = re.compile(r'_?[A-Z][a-zA-Z0-9]*$')
 
 
+# Node types which may contain class methods
+METHOD_CONTAINER_NODES = {ast.If, ast.While, ast.For, ast.With} | (
+    {ast.TryExcept, ast.TryFinally} if sys.version_info[0] == 2 else {ast.Try})
+
+
 if sys.version_info[0] < 3:
     def _unpack_args(args):
         ret = []
@@ -184,9 +189,15 @@ class NamingChecker(object):
         # If this class inherits from `type`, it's a metaclass, and we'll
         # consider all of it's methods to be classmethods.
         ismetaclass = any(name for name in cls_bases if name.id == 'type')
+        self.set_function_nodes_types(
+            iter_child_nodes(cls_node), ismetaclass, late_decoration)
 
+    def set_function_nodes_types(self, nodes, ismetaclass, late_decoration):
         # iterate over all functions and tag them
-        for node in iter_child_nodes(cls_node):
+        for node in nodes:
+            if type(node) in METHOD_CONTAINER_NODES:
+                self.set_function_nodes_types(
+                    iter_child_nodes(node), ismetaclass, late_decoration)
             if not isinstance(node, ast.FunctionDef):
                 continue
             node.function_type = _FunctionType.METHOD

--- a/testsuite/N807.py
+++ b/testsuite/N807.py
@@ -7,6 +7,10 @@ class C2:
     if True:
         def __str__(self):
             return ''
+#: N807
+if True:
+    def __bad__():
+        pass
 #: Okay
 class C3:
     try:

--- a/testsuite/N807.py
+++ b/testsuite/N807.py
@@ -1,3 +1,22 @@
+#: Okay
+class C1:
+    def __str__(self):
+        return ''
+#: Okay
+class C2:
+    if True:
+        def __str__(self):
+            return ''
+#: Okay
+class C3:
+    try:
+        if True:
+            while True:
+                def __str__(self):
+                    return ''
+                break
+    except:
+        pass
 #: N807
 def __bad():
     pass

